### ISSUE TITLE
`_PyEval_SetTrace` is no longer available in C API since 3.13

### ISSUE THREAD

**Issue by dalazx** [python/cpython#127748]
# Bug report

### Bug description:

Hello,

Prior 3.13 it was possible to set a trace in a specific thread from another thread by leveraging `_PyEval_SetTrace`, with pure Python. Now one can either set a trace for the current thread or all the threads using `PyEval_SetTrace` or `PyEval_SetTraceAllThreads`.
The change:
https://github.com/python/cpython/pull/108433/files#diff-9e7eb7e81667528dfa3a35bbf32fe622d4721c3c71385ea0ea2250abbd15d545R21
Interestingly, the profiling counterpart seems to be still available in the C API.

The use case I am interested in is a dedicated monitoring thread connected to a distributed observability system. This thread itself should not be traced as it manages tracing for the rest of the process.

Would it be possible to reinstate `_PyEval_SetTrace` back to the private C API?

Also, the new `sys.monitoring` module seems to offer a more high-level interface for debuggers and profilers, but it is unclear whether it is possible to limit the events to a certain thread. (unless of course one checks every event which does not seem efficient).

Alternatively, would writing an extension to expose this function be a better approach?

I would greatly appreciate any comments or suggestions regarding this. Thank you!

cc @vstinner 

### CPython versions tested on:

3.13

### Operating systems tested on:

macOS

**Comment by ZeroIntensity** [comment]
Calling `_PyEval_SetTrace` on a non-attached thread state is probably a thread safety headache because the fields are protected by the GIL (or by thread-localness on the free-threaded build), so I'm not sure you want to do that. If you have the pointer to the thread state that you want to mess with, you should attach to it with `PyThreadState_Swap` and then just call `PyEval_SetTrace`. Something like this should work:

```c
static void
PyEval_SetTrace_ForThread(PyThreadState *tstate, Py_tracefunc func, PyObject *obj)
{
    // Attach to target thread
    PyThreadState *old_tstate = PyThreadState_Swap(tstate);
    PyEval_SetTrace(func, obj)
    // Restore old thread
    PyThreadState_Swap(old_tstate);
}
```

**Comment by gaogaotiantian** [comment]
I can answer the `sys.monitoring` part - no there isn't a mechanism to limit the events on a specific thread at this point. The fundamental of `sys.monitoring` is dynamic opcodes, which changes the code object directly so it's code object specific. Even the current `sys.settrace` and `sys.setprofile` is based on `sys.monitoring`, and they have an explicit check on thread-specific trace/profile funcs (the underlying `sys.monitoring` event will always be triggered).

**Comment by dalazx** [comment]
@ZeroIntensity thank you for your reply.

`_PyEval_SetTrace` does check the GIL:
```
/* The caller must hold the GIL */
assert(PyGILState_Check());
```

and prior 3.13 it seemed quite simple to manage this using ctypes:
```
def _settrace_thread(
    tstate: ctypes._Pointer[PyThreadState], arg: TraceFunction | None
) -> None:
    func: ctypes.c_void_p
    argument: t.Any
    if arg is not None:
        func = _c_trace_trampoline
        argument = arg
    else:
        func = ctypes.cast(0, ctypes.c_void_p)
        argument = PyObjectPtr()
    _PyEval_SetTrace(tstate, func, argument)
```

I presume for `PyThreadState_Swap` I'll need to have an extension though.

**Comment by dalazx** [comment]
@gaogaotiantian thanks. I'll study the current `settrace` function closely.

**Comment by ZeroIntensity** [comment]
`PyGILState_Check` isn't extensive. On the free-threaded build and with subintepreters, accessing another thread state is non-atomic.

> I presume for PyThreadState_Swap I'll need to have an extension though.

Yeah, changing the thread state between Python calls isn't safe. Is there any reason you can't use an extension? (In my experience, ctypes is significantly more painful than just using an extension.)

**Comment by dalazx** [comment]
> Is there any reason you can't use an extension?

I thought it might slightly complicate package distribution with platform-specific wheels etc.

**Comment by ZeroIntensity** [comment]
I guess it would, but ctypes will open up a new complicated can of worms for debugging. IMO, if we *really* want to use ctypes, the best way to proceed here is not to reinstate `_PyEval_SetTrace`, but to just add some sort of mechanism into ctypes for changing the thread state. I might play around with developing that later today if I have time.

**Comment by dalazx** [comment]
thanks @ZeroIntensity.

do you mean trying to set the fields below directly using ctypes? 
https://github.com/python/cpython/blob/5c89adf385aaaca97c2ee9074f8b1fda0f57ad26/Python/legacy_tracing.c#L577-L582

**Comment by ZeroIntensity** [comment]
Nah, I mean actually trying to change Python's thread state form ctypes. It's not possible right now because the eval loop relies on it, but in theory some sort of wrapper that changes the thread state before calling into C could work. If I can come up with a reference implementation, I'll bring it to DPO.
